src/faebryk/libs/picker/jlcpcb.py

```diff
@@ -0,0 +1,946 @@
+
+    async def _close_db(self):
+        await Tortoise.close_connections()
+        self.connected = False
+
+    def has_db(self) -> bool:
+        return self.db_path.is_dir() and self.db_file.is_file()
+
+    def is_db_up_to_date(
+        self, max_timediff: datetime.timedelta = datetime.timedelta(days=7)
+    ) -> bool:
+        if not self.has_db():
+            return False
+
+        return (
+            datetime.datetime.fromtimestamp(
+                self.db_file.stat().st_mtime, tz=datetime.timezone.utc
+            )
+            >= datetime.datetime.now(tz=datetime.timezone.utc) - max_timediff
+        )
+
+    def prompt_db_update(self, prompt: str = "Update JLCPCB database?") -> bool:
+        ans = input(prompt + " [Y/n]:").lower()
+        return ans == "y" or ans == ""
+
+    def download(
+        self,
+    ):
+        zip_file = self.db_path / Path("cache.zip")
+
+        if not self.db_path.is_dir():
+            os.makedirs(self.db_path)
+
+        wget.download(
+            "https://yaqwsx.github.io/jlcparts/data/cache.zip",
+            out=str(zip_file),
+        )
+        # TODO: use requrests and 7z from python? (py7zr) and auto calc number
+        # of files
+        for i in range(1, 50):
+            try:
+                wget.download(
+                    f"https://yaqwsx.github.io/jlcparts/data/cache.z{i:02d}",
+                    out=str(self.db_path / Path(f"cache.z{i:02d}")),
+                )
+            except HTTPError:
+                break
+        subprocess.run(["7z", "x", str(zip_file), f"-o{self.db_path}"])
+
+    @dataclass
+    class parameter_to_db_map:
+        param_name: str
+        attr_keys: list[str]
+        attr_tolerance_key: str | None = None
+        transform_fn: Callable[[str], Any] = lambda x: x
+
+    async def find_by_lcsc_pn(self, partnumber: str, qty: int = 100):
+        filter_query = Q(stock__gte=qty) & Q(id=partnumber.strip("C"))
+        res = await Component.filter(filter_query).order_by("-basic")
+        if len(res) != 1:
+            raise PickError(
+                f"Could not find exact match for LCSC PN {partnumber} with qty {qty}"
+            )
+        await self._attach_component_to_module(Module(), res[0], [], qty)
+
+    async def find_by_manufacturer_pn(self, partnumber: str, qty: int = 100):
+        filter_query = Q(stock__gte=qty) & Q(mfr__icontains=partnumber)
+        res = await Component.filter(filter_query).order_by("-basic")
+        if len(res) < 1:
+            raise PickError(
+                f"Could not find match for manufacturer PN {partnumber} with qty {qty}"
+            )
+        res = self._sort_results_by_basic_preferred_price(res, qty)[0]
+        await self._attach_component_to_module(Module(), res, [], qty)
+
+    async def get_manufacturer_from_id(self, manufacturer_id: int) -> str:
+        return (await Manufacturers.get(id=manufacturer_id)).name
+
+    async def find_resistor(self, cmp: F.Resistor, qty: int = 100):
```

Comment:
I would make a trait: can_query_jlcpcb_db
Then for now move all those find_ functions into a single file: libs/picker/jlcpcb/pickers.py
Add a helper: find_query_for_module # attaches the trait with a big switch case
Add a function: pick_module_by_query

Make sure to raise the proper Exceptions so that the higher level pickers can react:

```python
# in app
def pick_parts(module: Module):
    try:
        pick_module_by_query(module)
    except PickerErrorNotImplemented:
        ...
    if isinstance(F.LED):
        pick_module_by_params(...)
    
```

Reply:
I didn't really implement the find function you posted, lmk if you still think it would be nice